### ISSUE TITLE
Spike Loss functions

### ISSUE THREAD

**Issue by cglagovich** [bamsumit/slayerPytorch#60]
I'm noticing that the default spike number targets for N-MNIST (60 and 10) work great for the 300ms of data, but perform worse as you decrease the sample length. I'm wondering if anybody has been doing work on a more generalized spike loss function such as a softmax? Do we think softmax could work in general? I would be interested in implementing this if it seems useful.

Thanks!

**Comment by bamsumit** [comment]
You would want to scale your target spike count proportionally.

A simple implementation of `numSpikes` loss without the target spike region is:
```
output = torch.sum(output, dim=-1)
desired = true_count * target + false_count * (1-target)
loss = F.mse_loss(output.flatten(), desired.flatten())
```

You could of course parameterize it to spike_rate (a value between 0 and 1)
```
output = torch.mean(output, dim=-1)
desired = true_rate * target + false_rate * (1-target)
loss = F.mse_loss(output.flatten(), desired.flatten())
```

And you can also use softmax loss after the mean/sum reduction of the output in time dimension. I generally find that rate based loss work well than softmax when tuned properly.
